crates/sui-core/src/accumulators/mod.rs

```diff
@@ -0,0 +1,248 @@
+    // Create a zero value with the appropriate type for the accumulator value.
+    fn zero(value: &AccumulatorValue) -> Self {
+        match value {
+            AccumulatorValue::Integer(_) => Self::SumU128(0),
+            AccumulatorValue::IntegerTuple(_, _) => Self::SumU128U128(0, 0),
+        }
+    }
+
+    fn accumulate_into(&mut self, value: AccumulatorValue) {
+        match (self, value) {
+            (Self::SumU128(v1), AccumulatorValue::Integer(v2)) => *v1 += v2 as u128,
+            (Self::SumU128U128(v1, v2), AccumulatorValue::IntegerTuple(w1, w2)) => {
+                *v1 += w1 as u128;
+                *v2 += w2 as u128;
+            }
+            _ => {
+                fatal!("invalid merge");
+            }
+        }
+    }
+}
+
+// TODO(address-balances): This currently only creates a single accumulator update transaction.
+// To support multiple accumulator update transactions, we need to:
+// - have each transaction take the accumulator root as a "non-exclusive mutable" input
+// - each transaction writes out a set of fields that are disjoint from the others.
+// - a barrier transaction must be added to advance the version of the accumulator root object.
+//   The barrier transaction doesn't do any field writes. This is necessary in order to provide
+//   a consistent view of the system accumulator state. When the version of the accumulator
+//   root object is advanced, we know that all accumulator state updates prior to that version
+//   have been applied.
+pub fn create_accumulator_update_transactions(
+    epoch_store: &AuthorityPerEpochStore,
+    checkpoint_height: u64,
+    cache: Option<&dyn TransactionCacheRead>,
+    ckpt_effects: &[TransactionEffects],
+) -> Vec<TransactionKind> {
+    let epoch = epoch_store.epoch();
+    let Some(accumulator_root_obj_initial_shared_version) = epoch_store
+        .epoch_start_config()
+        .accumulator_root_obj_initial_shared_version()
+    else {
+        debug!("accumulator root object does not exist, skipping accumulator update");
+        return vec![];
+    };
+
+    struct Update {
+        merge: MergedValueIntermediate,
+        split: MergedValueIntermediate,
+    }
+
+    let mut updates = HashMap::<_, Update>::new();
+
+    let mut addresses = HashMap::<_, AccumulatorAddress>::new();
+
+    for effect in ckpt_effects {
+        let tx = effect.transaction_digest();
+        // TransactionEffectsAPI::accumulator_events() uses a linear scan of all
+        // object changes and allocates a new vector. In the common case (on validators),
+        // we still have still have the original vector in the writeback cache, so
+        // we can avoid the unnecessary work by just taking it from the cache.
+        let events = match cache.and_then(|c| c.take_accumulator_events(tx)) {
+            Some(events) => events,
+            None => effect.accumulator_events(),
+        };
+
+        for AccumulatorEvent {
+            accumulator_obj,
```

Comment:
Could you remind me why we needed `accumulator_obj` in `AccumulatorEvent`, instead of deriving it using the address?

Reply:
we don't, but this is not an on chain structure, so I don't think the small memory savings would be worth all the re-hashing.